hack/generator/pkg/codegen/pipeline/unroll_recursive_types.go

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package pipeline
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+)
+
+const UnrollRecursiveTypesStageID = "unrollRecursiveTypes"
+
+// UnrollRecursiveTypes finds types that reference themselves and "unrolls" the reference. So a type that looks like:
+// type Error struct {
+//     code    string
+//     message string
+//     errors []Error
+// }
+// gets unrolled to look like:
+// type Error struct {
+//     code    string
+//     message string
+//     errors []Error_Unrolled
+// }
+// where Error_Unrolled looks like:
+// type Error_Unrolled struct {
+//     code    string
+//     message string
+// }
+// This usually happens for Error types (CloudError and the like). The recursive references must be removed because
+// controller-tools doesn't support generating "references" (JSON $ref) so it can't support recursive types today.
+// See https://github.com/kubernetes-sigs/controller-tools/issues/489 for more information.
+// Unrolling these types, while required for controller-tools to function, means
+// THAT THE TYPES WE GENERATE DON'T EXACTLY CONFORM TO THE PUBLISHED API!
+// In practice, we think this is ok because the types that reference themselves only return a depth of 1.
+// If we were to unroll all loops (rather than just types that directly reference themselves like we're doing here) that
+// "in practice" observation may no longer hold, so we avoid doing it here (it's also more complicated to do that).
+// TODO: We could require you to name the type in config or code in order to get it unrolled? There are currently 43 types being unrolled by this
+func UnrollRecursiveTypes() Stage {
+	return MakeStage(
+		UnrollRecursiveTypesStageID,
+		"Unrolls directly recursive types, as they are not supported by controller-gen",
```

Comment:
super minor nit: To make the tense consistent with other pipeline stages.
```suggestion
		"Unroll directly recursive types since they are not supported by controller-gen",
```

Reply:
Fixed